Start writing docs (#73)
inviqa/harness-go: docs/harness-attributes.md

# Using harness attributes

The Go harness has several attributes that you can use in your own project to control how it behaves. This document outlines some of the most common attributes that you will need to use when developing an application with it.

> _NOTE: This document is a work-in-progress, and only a subset of all attributes have been documented. Please consider contributing._

## Required attributes

Some attribute values are required in order to create a working application. You will be asked to provide these whenever you create a project (`ws create-project ...`):

> * `module_name` - this will be the name of the Go module for your created application, e.g. `github.com/inviqa/repo-name`, it is stored in the generated `go.mod` file.

Reviewer: ```suggestion
* `module_name` - this will be the name of the Go module for your created application, e.g. `github.com/inviqa/repo-name`. This is stored in the generated `go.mod` file.
```

## Overriding attributes

The harness has sensible defaults for all other attributes, but you will often need to override some of these in order to get the
Go harness to work for your particular application. To override an attribute, define an `attributes` property in your `workspace.yml` after your project has been created. It may look something like this:

    workspace('my-go-app'):
      description: generated local workspace for ci-go-sample.
      harness: inviqa/go

    attributes:
      # attributes should be added under here

You can then start to define attribute overrides.

### `go` attributes

These attributes control how Go behaves in the built docker image. A full picture of the attributes available under this node are as follows:  

    go:
      version: 1.15
      environment:
        CGO_ENABLED: '0'
        GOPRIVATE: github.com/inviqa/*
      modules:
        after:
          steps: []
        before:
          steps: []

#### `version`

> This controls the Go version to use, in the format of `major.minor`. When you create a project the harness will use the latest Go version available (defined [here](/harness/attributes/common.yml#L56)). If you have to change this be sure to also change the Go version in your `go.mod` file to match.

Reviewer: Do we want to just link to the supported versions document here? We could add the link to the latest version in `common.yml` to that document. Then this could just read something like:

```suggestion
This controls the Go version to use, in the format of `major.minor`. See [here](supported-go-versions.md) for the versions we support.
```
